tests/pytests/functional/cache/test_cache.py

```diff
@@ -0,0 +1,526 @@
+            assert actual_thing is new_thing
+        else:
+            assert actual_thing is not new_thing
+        assert actual_thing == new_thing
+
+    with subtests.test("contains returns true if key in bank"):
+        assert cache.contains(bank=bank, key=good_key)
+
+    with subtests.test("contains returns true if bank exists and key is None"):
+        assert cache.contains(bank=bank, key=None)
+
+    with subtests.test(
+        "contains returns False when bank not in cache and/or key not in bank"
+    ):
+        assert not cache.contains(bank=bank, key=bad_key)
+        assert not cache.contains(bank="nonexistent", key=good_key)
+        assert not cache.contains(bank="nonexistent", key=bad_key)
+        assert not cache.contains(bank="nonexistent", key=None)
+
+    with subtests.test("flushing nonexistent key should not remove other keys"):
+        cache.flush(bank=bank, key=bad_key)
+        assert cache.contains(bank=bank, key=good_key)
+
+    with subtests.test(
+        "flushing existing key should not remove bank if no more keys exist"
+    ):
+        pytest.skip(
+            "This is impossible with redis. Should we make localfs behave the same way?"
+        )
+        cache.flush(bank=bank, key=good_key)
+        assert cache.contains(bank=bank)
+        assert cache.list(bank=bank) == []
+
+    with subtests.test(
+        "after existing key is flushed updated should not return a timestamp for that key"
+    ):
+        cache.store(bank=bank, key=good_key, data="fnord")
+        cache.flush(bank=bank, key=good_key)
+        timestamp = cache.updated(bank=bank, key=good_key)
+        assert timestamp is None
+
+    with subtests.test(
+        "after flushing bank containing a good key, updated should not return a timestamp for that key"
+    ):
+        cache.store(bank=bank, key=good_key, data="fnord")
+        cache.flush(bank=bank, key=None)
+        timestamp = cache.updated(bank=bank, key=good_key)
+        assert timestamp is None
+
+    with subtests.test("flushing bank with None as key should remove bank"):
+        cache.flush(bank=bank, key=None)
+        assert not cache.contains(bank=bank)
+
+    with subtests.test("Exception should happen when flushing None bank"):
+        # This bit is maybe an accidental API, but currently there is no
+        # protection at least with the localfs cache when bank is None. If
+        # bank is None we try to `os.path.normpath` the bank, which explodes
+        # and is at least the current behavior. If we want to change that
+        # this test should change. Or be removed altogether.
+        # TODO: this should actually not raise. Not sure if there's a test that we can do here... or just call the code which will fail if there's actually an exception. -W. Werner, 2021-09-28
+        pytest.skip(
+            "Skipping for now - etcd, redis, and mysql do not raise. Should ensure all backends behave consistently"
+        )
```

Comment:
This is another accidental API. It's only the localfs cache that raises but I'm pretty sure it should not.